sql: stale metadata in logic_test

The following test appears to have failed:

[#8075](https://circleci.com/gh/cockroachdb/cockroach/8075):

```
tefce914d  127.0.0.1:0  17:32:49.423069  24.686µs     ·meta descriptor lookup       kv/dist_sender.go:510
tefce914d  127.0.0.1:0  17:32:49.423116  11.248288ms  ·sending RPC                  kv/dist_sender.go:424
tefce914d  127.0.0.1:0  17:32:49.423267  0            ··sending to 127.0.0.1:42792  rpc/send.go:156

I0928 17:32:49.440227 855 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- FAIL: TestLogic (4.33s)
	logic_test.go:393: testdata/database:52: expected "database \"b\" does not exist", but found success
=== RUN   TestPrivilege
--- PASS: TestPrivilege (0.02s)
=== RUN   TestPrivilegeValidate
--- PASS: TestPrivilegeValidate (0.01s)
=== RUN   TestSystemPrivilegeValidate
--- PASS: TestSystemPrivilegeValidate (0.01s)
=== RUN   TestRenameTable
I0928 17:32:49.489689 855 rpc/clock_offset.go:154  monitoring cluster offset
I0928 17:32:49.490004 855 multiraft/multiraft.go:442  node 100000001 starting
--
--- PASS: TestValidateTableDesc (0.00s)
=== RUN   TestColumnTypeSQLString
--- PASS: TestColumnTypeSQLString (0.00s)
=== RUN   TestInitialKeys
--- PASS: TestInitialKeys (0.00s)
FAIL
FAIL	github.com/cockroachdb/cockroach/sql	11.099s
=== RUN   TestNoLinkForbidden
--- PASS: TestNoLinkForbidden (1.06s)
=== RUN   TestPlaceholders
I0928 17:32:44.969158 863 rpc/clock_offset.go:154  monitoring cluster offset
I0928 17:32:44.975928 863 multiraft/multiraft.go:442  node 100000001 starting
I0928 17:32:44.976829 863 raft/raft.go:406  group 1 100000001 became follower at term 5
I0928 17:32:44.977112 863 raft/raft.go:219  group 1 newRaft 100000001 [peers: [100000001], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0928 17:32:44.977485 863 raft/raft.go:485  group 1 100000001 is starting a new election at term 5
I0928 17:32:44.977669 863 raft/raft.go:419  group 1 100000001 became candidate at term 6
I0928 17:32:44.977836 863 raft/raft.go:468  group 1 100000001 received vote from 100000001 at term 6

```
Please assign, take a look and update the issue accordingly.

--- a/sql/descriptor.go
+++ b/sql/descriptor.go
@@ -21,13 +21,15 @@ import (
 	"errors"
 	"fmt"
 
+	"github.com/gogo/protobuf/proto"
+
 	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/config"
 	"github.com/cockroachdb/cockroach/keys"
 	"github.com/cockroachdb/cockroach/roachpb"
 	"github.com/cockroachdb/cockroach/sql/parser"
 	"github.com/cockroachdb/cockroach/sql/privilege"
 	"github.com/cockroachdb/cockroach/util"
-	"github.com/gogo/protobuf/proto"
 )
 
 var (
@@ -73,9 +75,9 @@ func (p *planner) checkPrivilege(descriptor descriptorProto, privilege privilege
 // createDescriptor takes a Table or Database descriptor and creates it
 // if needed, incrementing the descriptor counter.
 func (p *planner) createDescriptor(plainKey descriptorKey, descriptor descriptorProto, ifNotExists bool) error {
-	key := plainKey.Key()
-	// Check whether key exists.
-	gr, err := p.txn.Get(key)
+	idKey := plainKey.Key()
+	// Check whether idKey exists.
+	gr, err := p.txn.Get(idKey)
 	if err != nil {
 		return err
 	}
@@ -108,13 +110,22 @@ func (p *planner) createDescriptor(plainKey descriptorKey, descriptor descriptor
 	descKey := MakeDescMetadataKey(descriptor.GetID())
 
 	b := client.Batch{}
-	b.CPut(key, descriptor.GetID(), nil)
-	b.CPut(descKey, wrapDescriptor(descriptor), nil)
+	descID := descriptor.GetID()
+	descDesc := wrapDescriptor(descriptor)
+	b.CPut(idKey, descID, nil)
+	b.CPut(descKey, descDesc, nil)
+
+	p.testingVerifyMetadata = func(systemConfig config.SystemConfig) error {
+		if err := expectDescriptorID(systemConfig, idKey, descID); err != nil {
+			return err
+		}
+		return expectDescriptor(systemConfig, descKey, descDesc)
+	}
 
 	return p.txn.Run(&b)
 }
 
-// getDescriptor looks up the descriptor for `key`, validates it,
+// getDescriptor looks up the descriptor for `plainKey`, validates it,
 // and unmarshals it into `descriptor`.
 func (p *planner) getDescriptor(plainKey descriptorKey, descriptor descriptorProto) error {
 	gr, err := p.txn.Get(plainKey.Key())

--- a/sql/drop.go
+++ b/sql/drop.go
@@ -20,12 +20,15 @@ package sql
 import (
 	"fmt"
 
+	"github.com/gogo/protobuf/proto"
+
 	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/config"
 	"github.com/cockroachdb/cockroach/keys"
+	"github.com/cockroachdb/cockroach/roachpb"
 	"github.com/cockroachdb/cockroach/sql/parser"
 	"github.com/cockroachdb/cockroach/sql/privilege"
 	"github.com/cockroachdb/cockroach/util"
-	"github.com/gogo/protobuf/proto"
 )
 
 // DropDatabase drops a database.
@@ -80,11 +83,22 @@ func (p *planner) DropDatabase(n *parser.DropDatabase) (planNode, error) {
 		return nil, err
 	}
 
+	zoneKey := MakeZoneKey(dbDesc.ID)
+
 	b := &client.Batch{}
 	b.Del(descKey)
 	b.Del(nameKey)
 	// Delete the zone config entry for this database.
-	b.Del(MakeZoneKey(dbDesc.ID))
+	b.Del(zoneKey)
+
+	p.testingVerifyMetadata = func(systemConfig config.SystemConfig) error {
+		for _, key := range [...]roachpb.Key{descKey, nameKey, zoneKey} {
+			if err := expectDeleted(systemConfig, key); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
 
 	if err := p.txn.Run(b); err != nil {
 		return nil, err
@@ -199,12 +213,23 @@ func (p *planner) DropTable(n *parser.DropTable) (planNode, error) {
 			return nil, err
 		}
 
+		zoneKey := MakeZoneKey(tableDesc.ID)
+
 		// Delete table descriptor
 		b := &client.Batch{}
 		b.Del(descKey)
 		b.Del(nameKey)
 		// Delete the zone config entry for this table.
-		b.Del(MakeZoneKey(tableDesc.ID))
+		b.Del(zoneKey)
+
+		p.testingVerifyMetadata = func(systemConfig config.SystemConfig) error {
+			for _, key := range [...]roachpb.Key{descKey, nameKey, zoneKey} {
+				if err := expectDeleted(systemConfig, key); err != nil {
+					return err
+				}
+			}
+			return nil
+		}
 
 		if err := p.txn.Run(b); err != nil {
 			return nil, err

--- a/sql/executor.go
+++ b/sql/executor.go
@@ -25,17 +25,31 @@ import (
 	"sync"
 	"time"
 
+	"github.com/gogo/protobuf/proto"
+
 	"github.com/cockroachdb/cockroach/client"
 	"github.com/cockroachdb/cockroach/config"
 	"github.com/cockroachdb/cockroach/gossip"
 	"github.com/cockroachdb/cockroach/roachpb"
 	"github.com/cockroachdb/cockroach/sql/driver"
 	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/util"
 	"github.com/cockroachdb/cockroach/util/hlc"
-	"github.com/gogo/protobuf/proto"
 )
 
+var testingWaitForMetadata bool
+
+// TestingWaitForMetadata causes metadata-mutating operations to wait
+// for the new metadata to back-propagate through gossip.
+func TestingWaitForMetadata() func() {
+	testingWaitForMetadata = true
+	return func() {
+		testingWaitForMetadata = false
+	}
+}
+
 var errNoTransactionInProgress = errors.New("there is no transaction in progress")
+var errStaleMetadata = errors.New("metadata is still stale")
 var errTransactionAborted = errors.New("current transaction is aborted, commands ignored until end of transaction block")
 var errTransactionInProgress = errors.New("there is already a transaction in progress")
 
@@ -55,8 +69,9 @@ type Executor struct {
 	leaseMgr *LeaseManager
 
 	// System Config and mutex.
-	systemConfig   *config.SystemConfig
-	systemConfigMu sync.RWMutex
+	systemConfig     config.SystemConfig
+	systemConfigMu   sync.RWMutex
+	systemConfigCond *sync.Cond
 }
 
 // newExecutor creates an Executor and registers a callback on the
@@ -67,6 +82,7 @@ func newExecutor(db client.DB, gossip *gossip.Gossip, clock *hlc.Clock) *Executo
 		reCache:  parser.NewRegexpCache(512),
 		leaseMgr: NewLeaseManager(0, db, clock),
 	}
+	exec.systemConfigCond = sync.NewCond(&exec.systemConfigMu)
 	gossip.RegisterSystemConfigCallback(exec.updateSystemConfig)
 	return exec
 }
@@ -81,16 +97,18 @@ func (e *Executor) SetNodeID(nodeID roachpb.NodeID) {
 // updateSystemConfig is called whenever the system config gossip entry is updated.
 func (e *Executor) updateSystemConfig(cfg *config.SystemConfig) {
 	e.systemConfigMu.Lock()
-	defer e.systemConfigMu.Unlock()
-	e.systemConfig = cfg
+	e.systemConfig = *cfg
+	e.systemConfigCond.Broadcast()
+	e.systemConfigMu.Unlock()
 }
 
 // getSystemConfig returns a pointer to the latest system config. May be nil,
 // if the gossip callback has not run.
-func (e *Executor) getSystemConfig() *config.SystemConfig {
+func (e *Executor) getSystemConfig() config.SystemConfig {
 	e.systemConfigMu.RLock()
-	defer e.systemConfigMu.RUnlock()
-	return e.systemConfig
+	cfg := e.systemConfig
+	e.systemConfigMu.RUnlock()
+	return cfg
 }
 
 // Execute the statement(s) in the given request and return a response.
@@ -303,6 +321,22 @@ func (e *Executor) execStmt(stmt parser.Statement, planMaker *planner) (driver.R
 		return result, err
 	}
 
+	if testingWaitForMetadata {
+		// We might need to verify metadata. Lock the system config so that
+		// no gossip updates sneak in under us.
+		// This lock does not change semantics. Even outside of tests, the
+		// planner is initialized with a static systemConfig, so locking
+		// the Executor's systemConfig cannot change the semantics of the
+		// SQL operation being performed under lock.
+		//
+		// The case of a multi-request transaction is not handled here,
+		// because those transactions outlive the verification callback.
+		// This can be addressed when we move to a connection-oriented
+		// protocol and server-side transactions.
+		e.systemConfigCond.L.Lock()
+		defer e.systemConfigCond.L.Unlock()
+	}
+
 	// No transaction. Run the command as a retryable block in an
 	// auto-transaction.
 	err := e.db.Txn(func(txn *client.Txn) error {
@@ -312,6 +346,25 @@ func (e *Executor) execStmt(stmt parser.Statement, planMaker *planner) (driver.R
 		planMaker.resetTxn()
 		return err
 	})
+
+	if testingWaitForMetadata && err == nil {
+		if verify := planMaker.testingVerifyMetadata; verify != nil {
+			// In the case of a multi-statement request, avoid reusing this
+			// callback.
+			planMaker.testingVerifyMetadata = nil
+			for i := 0; ; i++ {
+				if verify(e.systemConfig) != nil {
+					e.systemConfigCond.Wait()
+				} else {
+					if i != 1 {
+						err = util.Errorf("expected %s to require one gossip update, but it required %d", stmt, i)
+					}
+					break
+				}
+			}
+		}
+	}
+
 	return result, err
 }
 

--- a/sql/logic_test.go
+++ b/sql/logic_test.go
@@ -39,6 +39,7 @@ import (
 	"github.com/cockroachdb/cockroach/config"
 	"github.com/cockroachdb/cockroach/security"
 	"github.com/cockroachdb/cockroach/server"
+	csql "github.com/cockroachdb/cockroach/sql"
 	"github.com/cockroachdb/cockroach/testutils"
 	"github.com/cockroachdb/cockroach/util/leaktest"
 )
@@ -507,6 +508,7 @@ func (t *logicTest) success(file string) {
 
 func TestLogic(t *testing.T) {
 	defer leaktest.AfterTest(t)
+	defer csql.TestingWaitForMetadata()()
 
 	// TODO(marc): splitting ranges at table boundaries causes
 	// a blocked task and won't drain. Investigate and fix.

--- a/sql/plan.go
+++ b/sql/plan.go
@@ -41,11 +41,13 @@ type planner struct {
 	evalCtx      parser.EvalContext
 	leases       map[ID]*LeaseState
 	leaseMgr     *LeaseManager
-	systemConfig *config.SystemConfig
+	systemConfig config.SystemConfig
 
 	// TODO(pmattis): This is a hack to force updating to the latest version of a
 	// lease after a schema change operation such as CREATE INDEX.
 	modifiedSchemas []schemaInfo
+
+	testingVerifyMetadata func(config.SystemConfig) error
 
 	parser                   parser.Parser
 	extractAggregatesVisitor extractAggregatesVisitor

--- a/sql/rename.go
+++ b/sql/rename.go
@@ -22,6 +22,7 @@ import (
 	"fmt"
 
 	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/config"
 	"github.com/cockroachdb/cockroach/roachpb"
 	"github.com/cockroachdb/cockroach/security"
 	"github.com/cockroachdb/cockroach/sql/parser"
@@ -65,16 +66,31 @@ func (p *planner) RenameDatabase(n *parser.RenameDatabase) (planNode, error) {
 		return nil, err
 	}
 
+	newKey := databaseKey{string(n.NewName)}.Key()
+	oldKey := databaseKey{string(n.Name)}.Key()
+	descID := dbDesc.GetID()
+	descDesc := wrapDescriptor(dbDesc)
+
 	b := client.Batch{}
-	b.CPut(databaseKey{string(n.NewName)}.Key(), dbDesc.GetID(), nil)
-	b.Put(descKey, wrapDescriptor(dbDesc))
-	b.Del(databaseKey{string(n.Name)}.Key())
+	b.CPut(newKey, descID, nil)
+	b.Put(descKey, descDesc)
+	b.Del(oldKey)
 
 	if err := p.txn.Run(&b); err != nil {
 		if _, ok := err.(*roachpb.ConditionFailedError); ok {
 			return nil, fmt.Errorf("the new database name %q already exists", string(n.NewName))
 		}
 		return nil, err
+	}
+
+	p.testingVerifyMetadata = func(systemConfig config.SystemConfig) error {
+		if err := expectDescriptorID(systemConfig, newKey, descID); err != nil {
+			return err
+		}
+		if err := expectDescriptor(systemConfig, descKey, descDesc); err != nil {
+			return err
+		}
+		return expectDeleted(systemConfig, oldKey)
 	}
 
 	return &valuesNode{}, nil
@@ -152,9 +168,12 @@ func (p *planner) RenameTable(n *parser.RenameTable) (planNode, error) {
 		return nil, err
 	}
 
+	descID := tableDesc.GetID()
+	descDesc := wrapDescriptor(tableDesc)
+
 	b := client.Batch{}
-	b.Put(descKey, wrapDescriptor(tableDesc))
-	b.CPut(newTbKey, tableDesc.GetID(), nil)
+	b.Put(descKey, descDesc)
+	b.CPut(newTbKey, descID, nil)
 	b.Del(tbKey)
 
 	if err := p.txn.Run(&b); err != nil {
@@ -162,6 +181,16 @@ func (p *planner) RenameTable(n *parser.RenameTable) (planNode, error) {
 			return nil, fmt.Errorf("table name %q already exists", n.NewName.Table())
 		}
 		return nil, err
+	}
+
+	p.testingVerifyMetadata = func(systemConfig config.SystemConfig) error {
+		if err := expectDescriptorID(systemConfig, newTbKey, descID); err != nil {
+			return err
+		}
+		if err := expectDescriptor(systemConfig, descKey, descDesc); err != nil {
+			return err
+		}
+		return expectDeleted(systemConfig, tbKey)
 	}
 
 	return &valuesNode{}, nil

--- a/sql/table.go
+++ b/sql/table.go
@@ -247,12 +247,13 @@ func (p *planner) getTableID(qname *parser.QualifiedName) (ID, error) {
 	// the usage of a recently renamed table, but that's a race that could occur
 	// anyways.
 	nameKey := tableKey{dbDesc.ID, qname.Table()}
-	if nameVal := p.systemConfig.GetValue(nameKey.Key()); nameVal != nil {
+	key := nameKey.Key()
+	if nameVal := p.systemConfig.GetValue(key); nameVal != nil {
 		id, err := nameVal.GetInt()
 		return ID(id), err
 	}
 
-	gr, err := p.txn.Get(nameKey.Key())
+	gr, err := p.txn.Get(key)
 	if err != nil {
 		return 0, err
 	}

--- a/sql/verify.go
+++ b/sql/verify.go
@@ -1,0 +1,62 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tamir Duberstein (tamird@gmail.com)
+
+package sql
+
+import (
+	"github.com/gogo/protobuf/proto"
+
+	"github.com/cockroachdb/cockroach/config"
+	"github.com/cockroachdb/cockroach/roachpb"
+)
+
+func expectDescriptorID(systemConfig config.SystemConfig, idKey roachpb.Key, id ID) error {
+	idValue := systemConfig.GetValue(idKey)
+	if idValue == nil {
+		return errStaleMetadata
+	}
+	cachedID, err := idValue.GetInt()
+	if err != nil {
+		return err
+	}
+	if ID(cachedID) != id {
+		return errStaleMetadata
+	}
+	return nil
+}
+
+func expectDescriptor(systemConfig config.SystemConfig, idKey roachpb.Key, desc *Descriptor) error {
+	descValue := systemConfig.GetValue(idKey)
+	if descValue == nil {
+		return errStaleMetadata
+	}
+	var cachedDesc Descriptor
+	if err := descValue.GetProto(&cachedDesc); err != nil {
+		return err
+	}
+	if !proto.Equal(&cachedDesc, desc) {
+		return errStaleMetadata
+	}
+	return nil
+}
+
+func expectDeleted(systemConfig config.SystemConfig, key roachpb.Key) error {
+	if systemConfig.GetValue(key) != nil {
+		return errStaleMetadata
+	}
+	return nil
+}

